### PR TITLE
Make it easier to use DataGrip w/ Synapse's schema

### PR DESCRIPTION
**Pull request by DMRobertson**
See https://github.com/matrix-org/synapse/blob/dmr/schema-snapshot-dir/synapse/storage/schema/latest_dump/README.md for a description.

I considered putting this in a top-level directory, but I think Rich once commented we had enough of those.

If we don't want this, we can drop the PR.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -124,7 +124,10 @@ include = [
     { path = "rust/src/**", format = "sdist" },
 ]
 exclude = [
-    { path = "synapse/*.so", format = "sdist"}
+    { path = "synapse/*.so", format = "sdist" },
+    # This contains dev-only symlinks to the latest schema dumps. We don't want people
+    # to use it when installing.
+    { path = "synapse/storage/schema/latest_schema" },
```
Comment (clokep) [inline pyproject.toml:130]: Would it make more sense to put this all in `contrib/datagrip` or something? That could contain the symlink and the README?

Comment (DMRobertson) [reply pyproject.toml:130]: That's a good idea!

File: contrib/datagrip/README.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# Schema symlinks
+
+This directory contains symlinks to the latest dump of the postgres full schema. This is useful to have, as it allows IDEs to understand our schema and provide autocomplete, linters, inspections, etc.
+
+I (dmr) use the DataGrip functionality in IntelliJ's paid-for products. It seems to only consider files called `*.sql` when defining a schema from DDL; `*.sql.postgres` wasn't detected for me.
```
Comment (clokep) [inline contrib/datagrip/README.md:5]:
```suggestion
When using the DataGrip functionality in IntelliJ's products. It seems to only consider files called `*.sql` when defining a schema from DDL; `*.sql.postgres` will be ignored.
```

What's the ramification of this?

File: scripts-dev/make_full_schema.sh [hunk 3]
```diff
@@ -293,4 +299,12 @@ pg_dump --format=plain --data-only --inserts --no-tablespaces --no-acl --no-owne
 pg_dump --format=plain --schema-only         --no-tablespaces --no-acl --no-owner "$POSTGRES_STATE_DB_NAME"  | cleanup_pg_schema  > "$OUTPUT_DIR/state/full_schemas/$SCHEMA_NUMBER/full.sql.postgres"
 pg_dump --format=plain --data-only --inserts --no-tablespaces --no-acl --no-owner "$POSTGRES_STATE_DB_NAME"  | cleanup_pg_schema >> "$OUTPUT_DIR/state/full_schemas/$SCHEMA_NUMBER/full.sql.postgres"
 
+if [ "$OUTPUT_DIR" == *synapse/storage/schema ]; then
+  echo "Updating contrib/datagrip symlinks..."
+  ln -sf "synapse/storage/schema/common/full_schemas/$SCHEMA_NUMBER/full.sql.postgres" "contrib/datagrip/common.sql"
+  ln -sf "synapse/storage/schema/main/full_schemas/$SCHEMA_NUMBER/full.sql.postgres"   "contrib/datagrip/main.sql"
+  ln -sf "synapse/storage/schema/state/full_schemas/$SCHEMA_NUMBER/full.sql.postgres"  "contrib/datagrip/state.sql"
```
Comment (clokep) [inline scripts-dev/make_full_schema.sh:306]: I think this needs updating?

File: pyproject.toml (hunk 1 above)
Comment (clokep) [reply pyproject.toml:130]: I think these changes can be backed out?

File: scripts-dev/make_full_schema.sh (hunk 3 above)
Comment (DMRobertson) [reply scripts-dev/make_full_schema.sh:306]: Err, yes! Sorry, should have marked this as a draft again.

Comment (DMRobertson) [reply scripts-dev/make_full_schema.sh:306]: Err, actually I think this is correct. (`ln -s TARGET SYMLINK_LOCATION`)

File: contrib/datagrip/README.md (hunk 2 above)
Comment (DMRobertson) [reply contrib/datagrip/README.md:5]: Not sure what you're getting at. Some context:

My first attempt to get this working was to point the IDE at the directories `synapse/storage/[logical db]/full_schemas/72` and let it scan those files there. Unfortunately it only seems to look for files ending in `.sql`; it doesn't read the `.sql.postgres` nor `.sql.sqlite` files at all. I couldn't find a way to point the IDE at specific files either.

Hence my approach here: make symlinks to the postgres schema dumps ending in `.sql`, and place the links in one directory. (I deliberately chose to ignore SQLite.)

Will update the comment to better explain this.

File: scripts-dev/make_full_schema.sh (hunk 3 above)
Comment (DMRobertson) [reply scripts-dev/make_full_schema.sh:306]: Oh no, it needs some `../../`

Comment (clokep) [reply scripts-dev/make_full_schema.sh:306]: Arg, I always flip `ln` arguments around. 😢

File: contrib/datagrip/README.md (hunk 2 above)
Comment (clokep) [reply contrib/datagrip/README.md:5]: Maybe something like:

```suggestion
When using the DataGrip functionality in IntelliJ's products. It seems to only consider files called `*.sql` when defining a schema from DDL; `*.sql.postgres` will be ignored. To get around this we symlink those files to ones ending in `.sql`.
```

File: scripts-dev/make_full_schema.sh (hunk 3 above)
Comment (DMRobertson) [reply scripts-dev/make_full_schema.sh]: I tested the latest version in d383945e46 and this seemed to work for me.

File: contrib/datagrip/README.md (hunk 2 above)
Comment (DMRobertson) [reply contrib/datagrip/README.md]: See 1f1b5b068a

File: contrib/datagrip/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+In particular, the DataGrip functionality in IntelliJ's products seems to only consider files called `*.sql` when defining a schema from DDL; `*.sql.postgres` will be ignored. To get around this we symlink those files to ones ending in `.sql`. We've chosen to ignore the `.sql.sqlite` schema dumps here, as they're not intended for production use (and are much quicker to test against).
+
+## Example
+![](datagrip-aware-of-schema.png)
+
+## Caveats
+
+- Doesn't include temporary tables created ad-hoc by Synapse.
+- Postgres only. IDEs will likely be confused by SQLite-specific queries.
+- Will not include migrations created after the latest schema dump.
```
Comment (clokep) [inline contrib/datagrip/README.md:14]: I guess we could symlink each of these individually? That seems like overkill though...

**Review by clokep — APPROVED** [review]

**Pull request merged**